keps/sig-api-machinery/3617-fine-grained-authz/README.md

```diff
@@ -0,0 +1,1162 @@
+number of new authz checks__, certainly not more than O(log(number of fields))
+and hopefully even smaller than that.
+
+__This KEP will describe the places within API server where the new verbs will
+be checked.__ We are not modifying the authz system itself; we are making
+additional calls to it, of newly defined verbs.
+
+#### Schema Specification and Requirements
+
+First, note that many of the fields we know we wish to protect exist in
+__built-ins and custom resources (CRs); we'll support both__.
+
+There is already precedent for specifying attributes of fields directly in the
+schema. See server-side apply tags, CEL validations, documentation, etc. We
+propose a similar mechanism for API authors to state required special verbs.
+This will permit us to put those verbs in the documentation to make it __easy
+for cluster admins to discover__.
+
+The system for specifying custom verbs needs to permit:
+
+* Verbs which cover child fields; e.g. we need to be able to grant "all of
+  spec" instead of a (possibly unenumerable) list of verbs for each individual
+  field. We need this to keep the number of checks low.
+* Verbs which cover specific leaf fields, for example all labels in .metadata.labels.
+* Verbs which are parameterized in some way, examples given earlier. The schema
+  needs to state how the verb is parameterized.
+* If a field name changes between versions of an object, the permission verb
+  should not. Otherwise it will be extremely laborious to configure the system
+  to ensure permissions don't change when objects are accessed via different
+  versions.
+* Fields representing the same concept in *different* objects should share the
+  *same* permission. The authz system already has a spot for resource type,
+  which can be specific or wildcard based. We should not encode resource type
+  into the verb also.
+* Because the permissions are distinct from fields, we should do our best to
+  make the permissions (verb strings) not confusable for field names or paths.
+
+The exact permissions are going to be a shared responsibility between extension
+authors and in-tree Kubernetes developers, since anyone can author a CRD.
+
+As part of this KEP, we will update the schemas to supply permission verbs for
+the fields listed above. We will document the patterns used so that extension
+authors can make their own permission verbs which won't conflict with each other
+or with in-tree verbs. We may not know exactly what the best practices are until
+we have an alpha or beta behind us, hence we're not ready to specify that in
+this KEP yet.
+
+(The exact permission system will be described below because we're not done with
+the requirements yet, keep reading!)
+
+#### Uncovered verbs
+
+There's one more piece of information we need to satisfy our second goal, and it
+can't go in the schema because it may vary from cluster to cluster.
+
+This concept is whether a permission verb is "covered" by enclosing permissions,
```

Comment:
I'm not in love with "covered" and "unconvered" as terms - are they standard concepts in authz realm?

If not, maybe we can solicit ideas for names.  e.g. Inherited vs Explicit or something